src/System.Memory/src/System/SpanExtensions.cs

```diff
@@ -274,5 +274,116 @@ public static void CopyTo<T>(this T[] array, Span<T> destination)
+        }
+
+        /// <summary>
+        /// Determines whether two sequences overlap in memory.
+        /// </summary>
+        /// <remarks>
+        /// Two sequences overlap if they have positions in common and neither is empty.
+        /// Empty sequences do not overlap with any other sequence.
+        /// </remarks>
+        public static bool Overlaps<T>(this ReadOnlySpan<T> first, ReadOnlySpan<T> second)
+        {
+            if (first.IsEmpty || second.IsEmpty)
+            {
+                return false;
+            }
+
+            IntPtr byteOffset = Unsafe.ByteOffset(
+                ref first.DangerousGetPinnableReference(),
+                ref second.DangerousGetPinnableReference());
+
+            if (Unsafe.SizeOf<IntPtr>() == sizeof(int))
+            {
+                return (uint)byteOffset < (uint)(first.Length * Unsafe.SizeOf<T>()) ||
+                       (uint)byteOffset > (uint)-(second.Length * Unsafe.SizeOf<T>());
+            }
+            else
+            {
+                return (ulong)byteOffset < (ulong)((long)first.Length * Unsafe.SizeOf<T>()) ||
+                       (ulong)byteOffset > (ulong)-((long)second.Length * Unsafe.SizeOf<T>());
+            }
+        }
+
+        /// <summary>
+        /// Determines whether two sequences overlap in memory and outputs the element offset.
+        /// </summary>
+        /// <remarks>
+        /// Two sequences overlap if they have positions in common and neither is empty.
+        /// Empty sequences do not overlap with any other sequence. If the sequences
+        /// overlap, `elementOffset` is the number of elements by which the second sequence
+        /// is offset from the first sequence (i.e., second minus first). The number is
+        /// rounded away from zero if it is not a whole number, which can happen when
+        /// casting a sequence of a smaller type to a sequence of a larger type. If the
+        /// sequences do not overlap, `elementOffset` is meaningless and arbitrarily set to
+        /// zero.
+        /// </remarks>
+        public static bool Overlaps<T>(this ReadOnlySpan<T> first, ReadOnlySpan<T> second, out int elementOffset)
+        {
+            if (first.IsEmpty || second.IsEmpty)
+            {
+                elementOffset = 0;
+                return false;
+            }
+
+            IntPtr byteOffset = Unsafe.ByteOffset(
+                ref first.DangerousGetPinnableReference(),
+                ref second.DangerousGetPinnableReference());
+
+            if (Unsafe.SizeOf<IntPtr>() == sizeof(int))
+            {
+                if ((uint)byteOffset < (uint)(first.Length * Unsafe.SizeOf<T>()))
+                {
+                    elementOffset = ((int)byteOffset + (Unsafe.SizeOf<T>() - 1)) / Unsafe.SizeOf<T>();
+                    return true;
+                }
+                else if ((uint)byteOffset > (uint)-(second.Length * Unsafe.SizeOf<T>()))
```

Comment:
nit: at first glance, this awkward to read, can we explicitly multiply by negative 1?
`else if ((uint)byteOffset > (uint)(-1 * (second.Length * Unsafe.SizeOf<T>())))`

Also, can you please add a comment explaining why we negate this value?

Reply:
I don't think that writing `-1 * x` is in any way better than writing `-x` if you want to negate `x`.

The details of the calculations are illustrated in https://github.com/dotnet/corefx/pull/18731#issuecomment-296856781. I guess this question will come up every time somebody looks at the code. Should I add the illustration and explanation to the source code? (And perhaps move the Overlaps methods to a separate file to avoid cluttering?)

Reply:
> I don't think that writing `-1 * x` is in any way better than writing `-x` if you want to negate `x`.

In general, I agree. However, the type cast to (uint) makes it look strange. Although, it is just an opinion (which I don't feel strongly about).

I think that illustration in-line would be great! Same file should be OK.